lookout/style/format/rules.py

```diff
@@ -165,10 +171,30 @@ def _get_composite(feature_extractor: FeatureExtractor, labels: Tuple[int, ...])
         feature_extractor.labels_to_class_sequences.append(labels)
         return len(feature_extractor.labels_to_class_sequences) - 1
 
-    def harmonize_quotes(self, y_pred: numpy.ndarray,
-                         vnodes_y: Sequence[VirtualNode], vnodes: Sequence[VirtualNode],
-                         winners: numpy.ndarray, feature_extractor: FeatureExtractor,
-                         ) -> Tuple[numpy.ndarray, numpy.ndarray, "Rules"]:
+    def _group_quote_predictions(
+            self, vnodes_y: Sequence[VirtualNode], vnodes: Sequence[VirtualNode],
+            ) -> Mapping[int, Optional[Tuple[VirtualNode, VirtualNode, VirtualNode]]]:
+        quotes_classes = frozenset((CLASS_INDEX[CLS_DOUBLE_QUOTE], CLASS_INDEX[CLS_SINGLE_QUOTE]))
+        y_indices = {id(vnode): i for i, vnode in enumerate(vnodes_y)}
+        grouped_predictions = OrderedDict()
+        for vnode1, vnode2, vnode3 in zip(vnodes, islice(vnodes, 1, None),
+                                          islice(vnodes, 2, None)):
+            if (id(vnode1) not in y_indices or id(vnode3) not in y_indices or vnode2.node is None
+                    or vnode1.y[-1] not in quotes_classes or vnode3.y[0] != vnode1.y[-1]):
+                continue
+            vnode2_roles = frozenset(role_name(role_id) for role_id in vnode2.node.roles)
+            if "STRING" in vnode2_roles:
+                grouped_predictions[id(vnode1)] = vnode1, vnode2, vnode3
+                grouped_predictions[id(vnode3)] = None
+        return grouped_predictions
```

Comment:
Maybe it is better to return one index for `vnodes` sequence instead of 3 elements from it? The index of `vnode2`. You can be always sure that there are quotes tokes on the right and left.

Reply:
This is basically an iterator. Changing it to only return the index is akin to only implementing half of the iterator's logic for me. I'd rather return a triple to implement the full iterator logic